### ISSUE TITLE
optionally index only embedded paths or threads

### ISSUE THREAD

**Issue by ekg** [vgteam/vg#268]
I think it's always going to be better to employ haplotypes in our prior model than ignore them. For local alignment, we're more robust to recombination; but at the global scale it really increases the possible genome states and this is likely to hurt us when we work with large graphs. We've figured out how to work with graphs without haplotype information, but that doesn't mean it's the best way to go.

It should be possible to build the GCSA2 index using only embedded paths in the graph. The input to GCSA2 needs to be a de Bruijn graph with a single source and sink node. Walking the kmers of the embedded paths should be enough to derive the input to GCSA2.

I'd like to be ready to do this when we have full gPBWT integration. Rather than waiting, I can start testing on the MHC region, or any other graph which we assemble with `vg msga`.

I think the work required to make this happen is rather minimal. I need to write a function which iterates over the kmers in the paths of the graph, and another needs to optionally use it to get the kmers for GCSA2.

With a bit more coding we should be able to do the same thing using the gPBWT threads.

Possibly of interest to @adamnovak @benedictpaten @richarddurbin @jltsiren

**Comment by richarddurbin** [comment]
I have always been a fan of the GCSA2 index only covering sequences seen in embedded paths.  So I like this, and I see that gPBWT
should allow you to do this easily.  It would be good to get gPBWT for the 1000 Genomes haplotypes.

I looked at the gPBWT manuscript, at least the conceptual part at the front, and congratulations, I think the idea is correct and powerful.
I notice it uses the term “thread” rather than “walk” or “path”.  It would be good to know why and  standardise on some set of usage.
Also, I am not convinced that it needs sides.  I have been working on an initial draft of the vg manuscript with Erik, and we have homed
in on a node having two strands, one in each direction (the sequence of the reverse one being the reverse-complement of the sequence
of the forward one).  Then the gPBWT should be defined over strands not sides of nodes.  The advantage of this is that you only go through
one strand per node, not two sides, which is much more natural.  It would be good also to all agree on this nomenclature for vg-associated
graphs. 

Richard

> On 19 Mar 2016, at 16:22, Erik Garrison <notifications@github.com> wrote:
> 
> I think it's always going to be better to employ haplotypes in our prior model than ignore them. For local alignment, we're more robust to recombination; but at the global scale it really increases the possible genome states and this is likely to hurt us when we work with large graphs. We've figured out how to work with graphs without haplotype information, but that doesn't mean it's the best way to go.
> 
> It should be possible to build the GCSA2 index using only embedded paths in the graph. The input to GCSA2 needs to be a de Bruijn graph with a single source and sink node. Walking the kmers of the embedded paths should be enough to derive the input to GCSA2.
> 
> I'd like to be ready to do this when we have full gPBWT integration. Rather than waiting, I can start testing on the MHC region, or any other graph which we assemble with vg msga.
> 
> I think the work required to make this happen is rather minimal. I need to write a function which iterates over the kmers in the paths of the graph, and another needs to optionally use it to get the kmers for GCSA2.
> 
> With a bit more coding we should be able to do the same thing using the gPBWT threads.
> 
> Possibly of interest to @adamnovak <https://github.com/adamnovak> @benedictpaten <https://github.com/benedictpaten> @richarddurbin <https://github.com/richarddurbin> @jltsiren <https://github.com/jltsiren>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub <https://github.com/vgteam/vg/issues/268>




-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Comment by ekg** [comment]
Even in vg's API, we are using several concepts to refer to the nodes, which causes a lot of confusion. First there is the `Node*`, which is the backing storage for the node's label and id. Then we have `NodeTraversal`s, which I think are closest to "node strands": they are a `Node*` plus a boolean to say if we are backward. Finally, we have `NodeSide`s, which is something of a 90 degree rotation of `NodeTraversal`. It helps us talk about edges in relation to both strands of the node, which seemed to make sense in some places and awkward in others.

I would be happy to clarify all of this, but I am not sure if it's immediately necessary. I think the protobuf definitions should be extended to include a stranded node concept. From there it's a bit of hacking to switch everything into a single representation. It's not clear to me if we need to standardize on just `NodeStrand` or if we should continue to use both. I ran into this today while implementing the path-only indexing and needed to think for a minute about which to use.

As for the "thread" vs "path", I think that both of them have utility but we need to specify what they mean. In my understanding, "thread" is meant to be what we started calling an "embedded path", or a path that perfectly matches the graph. Right now I'd be most inclined to keep both around, and we were talking about implementing a new class to manage threads in vg. I'm waiting to see how @adamnovak implements the VCF thread extraction for the gPBWT tests. Also, we can use "thread" in the vg paper in place of the more cumbersome "fully embedded path".
